$\huge{solution \ \ of  \ \ 1.1 \ \ and   \ \ 1.2} $

let $x_{i \ j}$ be the indicator variable

 $x_{i \ j}$=1 if $i^{th}$ factory located at $j^{th}$ location
 
  $x_{i \ j} $= 0 elsewhere

let i k for factory i.e $ \ \ $ i,j ∈ {1,2,............n}

let j l  for location i.e$\ \ \ $ k,m ∈ {1,2,.............n}

let $d_{i \ k}$ be demand from $i^{th} \ \ factory \ \ to \ \ k^{th}$ factory

and let $u_{j \ l}$ be Unit costs of transportation from $j^{th}$ location to $l^{th}$ location

objective fuction is $ \sum_{i=1}^{i=n} \sum_{j=1}^{j=n} \sum_{k=1}^{k=n} \sum_{l=1}^{l=n} x_{i \ j} \ x_{k \ l} d_{i \ k} u_{j \ l}$

i.e if  $i^{th}$ factory located at $j^{th}$ location and $k^{th}$ factory located at $l^{th}$ location then demand from $i^{th} \ \ factory \ \ to \ \ k^{th}$ factory is $d_{i \ k}$ and Unit costs of transportation from $j^{th}$ location to $l^{th}$ location is $u_{j \ l}$ and we have to minimize the total cost

subjected to $ \sum_{j=1}^{j=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀i$

i.e each factory can be located max at a location

$ \sum_{i=1}^{i=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀j$

i.e each location has max 1 factory 

now our model is 

min $ \sum_{i=1}^{i=n} \sum_{j=1}^{j=n} \sum_{k=1}^{k=n} \sum_{l=1}^{l=n} x_{i \ j} \ x_{k \ l} d_{i \ k} u_{j \ l}$

subjected to $ \sum_{j=1}^{j=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀i$

$ \sum_{i=1}^{i=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀j$

i,k ∈ {1,2,..............n}

j,l ∈ {1,2,..................n}

$x_{i \ j} \ \ \in$ {0,1}

$x_{k \ l} \ \ \in$ {0,1}  

since we have to solve it by MILP so we have to convert it into MILP for that i take for first model  $y_{i \ j \ k \ l} = x_{i \ j} * x_{k \ l}$

min $ \sum_{i=1}^{i=n} \sum_{j=1}^{j=n} \sum_{k=1}^{k=n} \sum_{l=1}^{l=n} y_{i \ j \ k \ l} d_{i \ k} u_{j \ l}$

$y_{i \ j \ k \ l} \in$ {0,1}

$ \sum_{j=1}^{j=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀i$

$ \sum_{i=1}^{i=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀j$

i,k ∈ {1,2,..............n}

j,l ∈ {1,2,..................n}

$x_{i \ j} \ \ \in$ {0,1}


$y_{i \ j \ k \ l} \le x_{i \ j}  \ \ \  \ \forall i \ j \ k \ l$

$y_{i \ j \ k \ l} \le x_{k \ l} \ \ \ \  \forall i \ j \ k \ l$

  $y_{i \ j \ k \ l} \ge x_{k \ l} + x_{i \ j} -1 \ \ \ \  \forall i \ j \ k \ l$

  but it is hard to solve using pyomo since in that problem we have $ n^4+n^2$ binary variable and $3*n^4+2n$ constrains it takes too much time (pyomo1 model)

  again i tried to solve by another way of linearization for that i found a paper of LINEARIZING QUADRATIC ASSIGNMENT PROBLEM by Surya Prakash Singh published in april 2008

  he solved that problem by taking 
  
  $y_{i \ j \ k \ l} = x_{i \ j} * x_{k \ l}$

  min $ \sum_{i=1}^{i=n} \sum_{j=1}^{j=n} \sum_{k=1}^{k=n} \sum_{l=1}^{l=n} y_{i \ j \ k \ l} d_{i \ k} u_{j \ l}$

$y_{i \ j \ k \ l} \in$ {0,1}

$ \sum_{j=1}^{j=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀i$

$ \sum_{i=1}^{i=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀j$

$x_{i \ j} \ \ \in$ {0,1}

$y_{i \ j \ k \ l} + M(x_{k \ l} + x_{i \ j} -2) \le 1 \ \ \ \  \forall i \ j \ k \ l$

$y_{i \ j \ k \ l} - M(x_{k \ l} + x_{i \ j} -2) \ge 1 \ \ \ \  \forall i \ j \ k \ l$

in that problem we have $ n^4+n^2$ binary variable and $2*n^4 + 2*n$ constrains it also  takes too much time (our pyomo3 model)

in that paper he also gives relaxation and solved by 

  $y_{i \ j \ k \ l} = x_{i \ j} * x_{k \ l}$

  min $ \sum_{i=1}^{i=n} \sum_{j=1}^{j=n} \sum_{k=1}^{k=n} \sum_{l=1}^{l=n} y_{i \ j \ k \ l} d_{i \ k} u_{j \ l}$

$y_{i \ j \ k \ l} \in$ {0,1}

$ \sum_{j=1}^{j=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀i$

$ \sum_{i=1}^{i=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀j$

$x_{i \ j} \ \ \in$ {0,1}


$y_{i \ j \ k \ l} - M(x_{k \ l} + x_{i \ j} -2) \ge 1 \ \ \ \  \forall i \ j \ k \ l$

in that problem we have $ n^4+n^2$ binary variable and $n^4 + 2*n$ constrains it also  takes too much time (our pyomo 2 model)


an another linearization is also called adams and johnsons linearization

that is 

min $ \sum_{i=1}^{i=n} \sum_{j=1}^{j=n} \sum_{k=1}^{k=n} \sum_{l=1}^{l=n} y_{i \ j \ k \ l} d_{i \ k} u_{j \ l}$

$ \sum_{i=1}^{i=n} y_{i \ j \ k \ l}=x_{k \ l} \ \ \ \  \forall j \ k \ l $

$ \sum_{j=1}^{j=n} y_{i \ j \ k \ l}=x_{k \ l} \ \ \ \  \forall i \ k \ l $

$ y_{i \ j \ k \ l}= y_{k \ l \ i \ j }  \ \ \ \ \ \forall i \ j \ k \ l$


$ y_{i \ j \ k \ l} \ge 0$


$ \sum_{j=1}^{j=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀i$

$ \sum_{i=1}^{i=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀j$

$x_{i \ j} \in $ {0,1}

in that problem we have $ n^2$ binary variable $n^4$ nonnegativereal variable  and $n^4 + 2*n^3 + 2*n$ constrains

(our pyomo4 model)





In [1]:
import numpy as np
import pandas as pd
!pip install -q pyomo
from pyomo.environ import *
!apt-get install -y -qq coinor-cbc
opt_cbc=SolverFactory('cbc')

In [2]:
def pyomo1(demand,unit_cost,n):
  global model
  model=ConcreteModel()
  model.y=Var(np.arange(n),np.arange(n),np.arange(n),np.arange(n),domain=Binary)
  model.x=Var(np.arange(n),np.arange(n),domain=Binary)
  model.constraints=ConstraintList()
  model.objective=Objective(expr=sum(demand[i,k]*unit_cost[j,l]*model.y[i,j,k,l] for i in np.arange(n) for j in np.arange(n) for k in np.arange(n) for l in np.arange(n)),sense=minimize)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[i,j] for j in range(n))==1)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[j,i] for j in range(n))==1)
    for i in range(n):
      for k in range(n):
        for l in range(n):
          for j in range(n):
             model.constraints.add(model.y[i,j,k,l]<=model.x[i,j])
    for i in range(n):
      for k in range(n):
        for l in range(n):
          for j in range(n):
             model.constraints.add(model.y[i,j,k,l]<=model.x[k,l])
  for i in range(n):
    for k in range(n):
      for l in range(n):
        for j in range(n):
           model.constraints.add(model.y[i,j,k,l]>=model.x[k,l]+model.x[i,j]-1)

In [3]:
def pyomo2(demand,unit_cost,n):
  global model
  model=ConcreteModel()
  model.y=Var(np.arange(n),np.arange(n),np.arange(n),np.arange(n),domain=Binary)
  model.x=Var(np.arange(n),np.arange(n),domain=Binary)
  model.constraints=ConstraintList()
  model.objective=Objective(expr=sum(demand[i,k]*unit_cost[j,l]*model.y[i,j,k,l] for i in np.arange(n) for j in np.arange(n) for k in np.arange(n) for l in np.arange(n)),sense=minimize)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[i,j] for j in range(n))==1)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[j,i] for j in range(n))==1)
  for i in range(n):
    for k in range(n):
      for l in range(n):
        for j in range(n):
          model.constraints.add(model.y[i,j,k,l] - 6000*(model.x[i,j] + model.x[k,l]-2) >=1)


In [4]:
def pyomo3(demand,unit_cost,n):
  global model
  model=ConcreteModel()
  model.y=Var(np.arange(n),np.arange(n),np.arange(n),np.arange(n),domain=Binary)
  model.x=Var(np.arange(n),np.arange(n),domain=Binary)
  model.constraints=ConstraintList()
  model.objective=Objective(expr=sum(demand[i,k]*unit_cost[j,l]*model.y[i,j,k,l] for i in np.arange(n) for j in np.arange(n) for k in np.arange(n) for l in np.arange(n)),sense=minimize)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[i,j] for j in range(n))==1)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[j,i] for j in range(n))==1)
  for i in range(n):
    for k in range(n):
      for l in range(n):
        for j in range(n):
          model.constraints.add(model.y[i,j,k,l] - 6000*(model.x[i,j] + model.x[k,l]-2) >=1)
  for i in range(n):
    for k in range(n):
      for l in range(n):
        for j in range(n):
          model.constraints.add(model.y[i,j,k,l] + 6000*(model.x[i,j] + model.x[k,l]-2) <=1)
  

In [5]:
def pyomo4(demand,unit_cost,n):
  global model4
  model4=ConcreteModel()
  model4.y=Var(np.arange(n),np.arange(n),np.arange(n),np.arange(n),domain=NonNegativeReals)
  model4.x=Var(np.arange(n),np.arange(n),domain=Binary)
  model4.objective=Objective(expr=sum(demand[i,k]*unit_cost[j,l]*model4.y[i,j,k,l] for i in np.arange(n) for j in np.arange(n) for k in np.arange(n) for l in np.arange(n)),sense=minimize)
  model4.constraints=ConstraintList()
  for j in range(n):
    for k in range(n):
      for l in range(n):
        model4.constraints.add(expr=sum(model4.y[i,j,k,l] for i in range(n))==model4.x[k,l])
  for i in range(n):
    model4.constraints.add(expr=sum(model4.x[i,j] for j in range(n))==1)
  for i in range(n):
    model4.constraints.add(expr=sum(model4.x[j,i] for j in range(n))==1)
  for i in range(n):
    for k in range(n):
      for l in range(n):
        model4.constraints.add(expr=sum(model4.y[i,j,k,l] for j in range(n))==model4.x[k,l])
  for i in range(n):
    for k in range(n):
      for l in range(n):
        for j in range(n):
          model4.constraints.add(model4.y[i,j,k,l]==model4.y[k,l,i,j])

our pyomo model5

min $ \sum_{i=1}^{i=n} \sum_{j=1}^{j=n} \sum_{k=1}^{k=n} \sum_{m=1}^{m=n} y_{i \ j \ k \ l} d_{i \ k} u_{j \ l}$

$y_{i \ j \ k \ l} - x_{i \ j}-x_{k \ l} \ge -1  \ \ \ \ \forall i \ j \ k \ l$


$y_{i \ j \ k \ l} \ge 0$

$ \sum_{j=1}^{j=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀i$

$ \sum_{i=1}^{i=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀j$

$x_{i \ j} \ \ \in$ {0,1}

here n^4 +2n constraints and n^4 real variable and n^2 binary variable

In [6]:
def pyomo5(demand,unit_cost,n):
  model=ConcreteModel()
  model.y=Var(np.arange(n),np.arange(n),np.arange(n),np.arange(n),domain=NonNegativeReals)
  model.x=Var(np.arange(n),np.arange(n),domain=Binary)
  model.constraints=ConstraintList()
  model.objective=Objective(expr=sum(demand[i,k]*unit_cost[j,l]*model.y[i,j,k,l] for i in np.arange(n) for j in np.arange(n) for k in np.arange(n) for l in np.arange(n)),sense=minimize)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[i,j] for j in range(n))==1)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[j,i] for j in range(n))==1)
  for i in range(n):
    for k in range(n):
      for l in range(n):
        for j in range(n):
           model.constraints.add(model.y[i,j,k,l] - model.x[i,j]-model.x[k,l] >=-1)

our pyomo model6

Lawler’s Linearization

min $ \sum_{i=1}^{i=n} \sum_{j=1}^{j=n} \sum_{k=1}^{k=n} \sum_{l=1}^{l=n} y_{i \ j \ k \ l} d_{i \ k} u_{j \ l}$

$\sum_{i=1}^{i=n} \sum_{j=1}^{j=n} \sum_{k=1}^{k=n} \sum_{l=1}^{l=n} y_{i \ j \ k \ l}=n^2$


$-2*y_{i \ j \ k \ l} + x_{i \ j}+x_{k \ l} \ge 0  \ \ \ \ \forall i \ j \ k \ l$


$y_{i \ j \ k \ l} \in $ {0,1}

$ \sum_{j=1}^{j=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀i$

$ \sum_{i=1}^{i=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀j$

$x_{i \ j} \ \ \in$ {0,1}

here $n^4 +n^2$ binary variable $n^4+2n+1 $ constraints

In [7]:
def pyomo6(demand,unit_cost,n):
  global model
  model=ConcreteModel()
  model.y=Var(np.arange(n),np.arange(n),np.arange(n),np.arange(n),domain=Binary)
  model.x=Var(np.arange(n),np.arange(n),domain=Binary)
  model.constraints=ConstraintList()
  model.objective=Objective(expr=sum(demand[i,k]*unit_cost[j,l]*model.y[i,j,k,l] for i in np.arange(n) for j in np.arange(n) for k in np.arange(n) for l in np.arange(n)),sense=minimize)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[i,j] for j in range(n))==1)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[j,i] for j in range(n))==1)
  for i in range(n):
    for k in range(n):
      for l in range(n):
        for j in range(n):
           model.constraints.add(-2*model.y[i,j,k,l] + model.x[i,j]+model.x[k,l] >=0)
    model.constraints.add(expr=sum(model.y[i,j,k,l] for i in np.arange(n) for j in np.arange(n) for k in np.arange(n) for l in np.arange(n))==n**2)

Frieze and Yadegar Linearization (pyomo7 model)


min $ \sum_{i=1}^{i=n} \sum_{j=1}^{j=n} \sum_{k=1}^{k=n} \sum_{l=1}^{l=n} y_{i \ j \ k \ l} d_{i \ k} u_{j \ l}$

$ \sum_{j=1}^{j=n} y_{i \ j \ k \ l}=x_{k \ l} \ \ \ \  \forall i \ k \ l $

$ \sum_{i=1}^{i=n} y_{i \ j \ k \ l}=x_{k \ l} \ \ \ \  \forall j \ k \ l $

$ \sum_{k=1}^{k=n} y_{i \ j \ k \ l}=x_{i \ j} \ \ \ \  \forall i \ j \ l $

$ \sum_{l=1}^{l=n} y_{i \ j \ k \ l}=x_{i \ j} \ \ \ \  \forall i \ j \ k $

$ y_{i \ j \ i \ j}=x_{i \ j} \forall i,j$

$0 \le y_{i \ j \ i \ j} \le 1$

$ \sum_{j=1}^{j=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀i$

$ \sum_{i=1}^{i=n} \  x_{i \ j} \ = 1 \ \ \ \ ∀j$

$x_{i \ j} \ \ \in$ {0,1}

i,j,k,l= {1,2,3,..................n}

here $4n^3+n^2+2n$ constraints $n^4 $ real variable and $n^2$ binary variable 

In [8]:
def pyomo7(demand,unit_cost,n):
  global model
  model=ConcreteModel()
  model.y=Var(np.arange(n),np.arange(n),np.arange(n),np.arange(n),domain=Reals)
  model.y.setlb(0)
  model.y.setub(1)
  model.x=Var(np.arange(n),np.arange(n),domain=Binary)
  model.constraints=ConstraintList()
  model.objective=Objective(expr=sum(demand[i,k]*unit_cost[j,l]*model.y[i,j,k,l] for i in np.arange(n) for j in np.arange(n) for k in np.arange(n) for l in np.arange(n)),sense=minimize)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[i,j] for j in range(n))==1)
  for i in range(n):
    model.constraints.add(expr=sum(model.x[j,i] for j in range(n))==1)
  for j in range(n):
    for k in range(n):
      for l in range(n):
        model.constraints.add(expr=sum(model.y[i,j,k,l] for i in range(n))==model.x[k,l])
  for i in range(n):
    for k in range(n):
      for l in range(n):
        model.constraints.add(expr=sum(model.y[i,j,k,l] for j in range(n))==model.x[k,l])
  for i in range(n):
    for j in range(n):
      for l in range(n):
        model.constraints.add(expr=sum(model.y[i,j,k,l] for k in range(n))==model.x[i,j])
  for i in range(n):
    for j in range(n):
      for k in range(n):
        model.constraints.add(expr=sum(model.y[i,j,k,l] for l in range(n))==model.x[i,j])
  for i in range(n):
    for j in range(n):
      model.constraints.add(model.y[i,j,i,j]==model.x[i,j])

In [9]:
demand=np.loadtxt("Lab7_ex1_q.txt",delimiter='   ')
unit_cost=np.loadtxt("Lab7_ex1_c.txt",delimiter='   ')

In [10]:
pyomo7(demand,unit_cost,8)

In [11]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
  # display solution
print('\nObjective = ', model.objective())

print(result.solver.time)

for i in range(8):
  for j in range(8):
    if model.x[i,j].value==1:
      print(i+1,'factory opened at ',j+1,'th location')

Solver status: ok
Solver termination condition: optimal

Objective =  14889.0
83.45413947105408
1 factory opened at  2 th location
2 factory opened at  8 th location
3 factory opened at  7 th location
4 factory opened at  6 th location
5 factory opened at  1 th location
6 factory opened at  3 th location
7 factory opened at  5 th location
8 factory opened at  4 th location


$\huge{solution \ \ of \ \ 1.9}$

Objective cost =  14889.0

1 factory opened at  2 th location

2 factory opened at  8 th location

3 factory opened at  7 th location

4 factory opened at  6 th location

5 factory opened at  1 th location

6 factory opened at  3 th location

7 factory opened at  5 th location

8 factory opened at  4 th location

In [12]:
pyomo4(demand,unit_cost,8)

In [13]:
result = opt_cbc.solve(model4)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
  # display solution
print('\nObjective = ', model4.objective())
print(result.solver.time)
for i in range(8):
  for j in range(8):
    if model4.x[i,j].value==1:
      print(i+1,'factory opened  at ',j+1,'th location')

Solver status: ok
Solver termination condition: optimal

Objective =  14889.0
716.9786298274994
1 factory opened  at  2 th location
2 factory opened  at  8 th location
3 factory opened  at  7 th location
4 factory opened  at  6 th location
5 factory opened  at  1 th location
6 factory opened  at  3 th location
7 factory opened  at  5 th location
8 factory opened  at  4 th location


In [14]:
pyomo4(demand,unit_cost,8)
model4.x.domain=NonNegativeReals
model4.x.setub(1)

In [15]:
result = opt_cbc.solve(model4)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print(result.solver.time)
  # display solution
print('\nObjective = ', model4.objective())
for i in range(8):
  for j in range(8):
    if model4.x[i,j].value!=0:
      print('x',i,j,'=',model4.x[i,j].value)

Solver status: ok
Solver termination condition: optimal
1.6354789733886719

Objective =  11790.875009541658
x 0 0 = 0.084983499
x 0 1 = 0.33367696
x 0 2 = 0.048316653
x 0 3 = 0.10039981
x 0 4 = 0.051444116
x 0 5 = 0.25047975
x 0 6 = 0.066562267
x 0 7 = 0.064136939
x 1 0 = 0.19073548
x 1 1 = 0.026261001
x 1 2 = 0.11864323
x 1 3 = 0.1917243
x 1 4 = 0.17399296
x 1 5 = 0.022830959
x 1 6 = 0.15180168
x 1 7 = 0.1240104
x 2 0 = 0.10388789
x 2 2 = 0.17341885
x 2 3 = 0.20991576
x 2 4 = 0.15765049
x 2 5 = 0.12777006
x 2 6 = 0.12970289
x 2 7 = 0.09765406
x 3 0 = 0.07739104
x 3 1 = 0.0017389244
x 3 2 = 0.16883651
x 3 3 = 0.041916945
x 3 4 = 0.22827876
x 3 5 = 0.13528148
x 3 6 = 0.11864323
x 3 7 = 0.22791312
x 4 0 = 0.20537663
x 4 1 = 0.14760491
x 4 2 = 0.15701317
x 4 3 = 0.043650923
x 4 4 = 0.038651246
x 4 5 = 0.0772711
x 4 6 = 0.15673067
x 4 7 = 0.17370134
x 5 0 = 0.022830959
x 5 2 = 0.19583209
x 5 3 = 0.019085987
x 5 4 = 0.26693
x 5 5 = 0.32892201
x 5 6 = 0.06176476
x 5 7 = 0.1046342
x 6 0 = 0.0

$\huge{solution \ \ of \ \ 1.10}$

Objective =  11790.875009541658

x 0 0 = 0.084983499

x 0 1 = 0.33367696

x 0 2 = 0.048316653

x 0 3 = 0.10039981

x 0 4 = 0.051444116

x 0 5 = 0.25047975

x 0 6 = 0.066562267

x 0 7 = 0.064136939

x 1 0 = 0.19073548

x 1 1 = 0.026261001

x 1 2 = 0.11864323

x 1 3 = 0.1917243

x 1 4 = 0.17399296

x 1 5 = 0.022830959

x 1 6 = 0.15180168

x 1 7 = 0.1240104

x 2 0 = 0.10388789

x 2 2 = 0.17341885

x 2 3 = 0.20991576

x 2 4 = 0.15765049

x 2 5 = 0.12777006

x 2 6 = 0.12970289

x 2 7 = 0.09765406

x 3 0 = 0.07739104

x 3 1 = 0.0017389244

x 3 2 = 0.16883651

x 3 3 = 0.041916945

x 3 4 = 0.22827876

x 3 5 = 0.13528148

x 3 6 = 0.11864323

x 3 7 = 0.22791312

x 4 0 = 0.20537663

x 4 1 = 0.14760491

x 4 2 = 0.15701317

x 4 3 = 0.043650923

x 4 4 = 0.038651246

x 4 5 = 0.0772711

x 4 6 = 0.15673067

x 4 7 = 0.17370134

x 5 0 = 0.022830959

x 5 2 = 0.19583209

x 5 3 = 0.019085987

x 5 4 = 0.26693

x 5 5 = 0.32892201

x 5 6 = 0.06176476

x 5 7 = 0.1046342

x 6 0 = 0.063140624

x 6 1 = 0.31854682

x 6 2 = 0.06176476

x 6 3 = 0.26544384

x 6 4 = 0.083052425

x 6 5 = 0.023820054

x 6 6 = 0.063140624

x 6 7 = 0.12109086

x 7 0 = 0.25165388


x 7 1 = 0.17217138

x 7 2 = 0.076174743

x 7 3 = 0.12786244

x 7 5 = 0.033624599

x 7 6 = 0.25165388

x 7 7 = 0.086859086

$\huge{solution \ \ of \ \ 1.11}$

optimal cost are not same and variable are not integer value now 

The MILP is bounded by some polyhedron with the additional constraint that only the integral points are feasible. The LP MODEL  is the same polyhedron where all points are feasible. Therefore, the first condition is satisfied as the set of all points in a polyhedron must logically include all integral points in the same polyhedron

now,if we  solve LP  and all variables are integral. Either (MILP) has the integrality property (where the optimal solution of the corresponding LP will always be integral, such as with LPs with totally unimodular matrices), or we were lucky and the optimal solution of MILP touches the edge of the polyhedron that bounds the feasible region

we can be certain that (for a minimisation problem), the optimal value to the corresponding MILP must be at least the optimal solution to LP, since the set of feasible points of MILP is contained inside that of LP, and because the function values of MILP at any arbitrary point is at least as great as that of LP at the same point.

i.e if soltion of LP is integer lies in the feasible reason of MILP then both solution are equal i.e feasible solution for the MILP is a feasible solution to the LP, and the reversed is not always so, i.e. a feasible solution for the LP is not necessarily a feasible solution for the MILP

In [16]:
'''pyomo1(demand,unit_cost,8)
model.y.domain=NonNegativeReals
model.y.setub(1)'''

'pyomo1(demand,unit_cost,8)\nmodel.y.domain=NonNegativeReals\nmodel.y.setub(1)'

In [17]:
'''result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
  # display solution
print('\nObjective = ', model.objective())
result.solver.time
print(result)'''

"result = opt_cbc.solve(model)\nprint('Solver status:', result.solver.status)\nprint('Solver termination condition:',result.solver.termination_condition)\n  # display solution\nprint('\nObjective = ', model.objective())\nresult.solver.time\nprint(result)"

In [18]:
#pyomo2(demand,unit_cost,8)

In [19]:
'''result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
  # display solution
print('\nObjective = ', model.objective())
result.solver.time
print(result)'''

"result = opt_cbc.solve(model)\nprint('Solver status:', result.solver.status)\nprint('Solver termination condition:',result.solver.termination_condition)\n  # display solution\nprint('\nObjective = ', model.objective())\nresult.solver.time\nprint(result)"

In [20]:
#pyomo3(demand,unit_cost,8)

In [21]:
'''result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
  # display solution
print('\nObjective = ', model.objective())
result.solver.time
print(result)'''

"result = opt_cbc.solve(model)\nprint('Solver status:', result.solver.status)\nprint('Solver termination condition:',result.solver.termination_condition)\n  # display solution\nprint('\nObjective = ', model.objective())\nresult.solver.time\nprint(result)"

In [22]:
'''pyomo2(demand,unit_cost,8)
model.y.domain=NonNegativeReals
model.y.setub(1)'''

'pyomo2(demand,unit_cost,8)\nmodel.y.domain=NonNegativeReals\nmodel.y.setub(1)'

In [23]:
'''result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
  # display solution
print('\nObjective = ', model.objective())
result.solver.time
print(result)'''

"result = opt_cbc.solve(model)\nprint('Solver status:', result.solver.status)\nprint('Solver termination condition:',result.solver.termination_condition)\n  # display solution\nprint('\nObjective = ', model.objective())\nresult.solver.time\nprint(result)"

In [24]:
'''pyomo3(demand,unit_cost,8)
model.y.domain=NonNegativeReals
model.y.setub(1)'''

'pyomo3(demand,unit_cost,8)\nmodel.y.domain=NonNegativeReals\nmodel.y.setub(1)'

In [25]:
'''result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
  # display solution
print('\nObjective = ', model.objective())
result.solver.time
print(result)'''

"result = opt_cbc.solve(model)\nprint('Solver status:', result.solver.status)\nprint('Solver termination condition:',result.solver.termination_condition)\n  # display solution\nprint('\nObjective = ', model.objective())\nresult.solver.time\nprint(result)"